clients/src/main/java/org/apache/kafka/common/config/SaslConfigs.java

```diff
@@ -202,7 +202,7 @@ public static void addClientSaslSupport(ConfigDef config) {
                 .define(SaslConfigs.SASL_LOGIN_REFRESH_WINDOW_JITTER, ConfigDef.Type.DOUBLE, SaslConfigs.DEFAULT_LOGIN_REFRESH_WINDOW_JITTER, Range.between(0.0, 0.25), ConfigDef.Importance.LOW, SaslConfigs.SASL_LOGIN_REFRESH_WINDOW_JITTER_DOC)
                 .define(SaslConfigs.SASL_LOGIN_REFRESH_MIN_PERIOD_SECONDS, ConfigDef.Type.SHORT, SaslConfigs.DEFAULT_LOGIN_REFRESH_MIN_PERIOD_SECONDS, Range.between(0, 900), ConfigDef.Importance.LOW, SaslConfigs.SASL_LOGIN_REFRESH_MIN_PERIOD_SECONDS_DOC)
                 .define(SaslConfigs.SASL_LOGIN_REFRESH_BUFFER_SECONDS, ConfigDef.Type.SHORT, SaslConfigs.DEFAULT_LOGIN_REFRESH_BUFFER_SECONDS, Range.between(0, 3600), ConfigDef.Importance.LOW, SaslConfigs.SASL_LOGIN_REFRESH_BUFFER_SECONDS_DOC)
-                .define(SaslConfigs.SASL_MECHANISM, ConfigDef.Type.STRING, SaslConfigs.DEFAULT_SASL_MECHANISM, ConfigDef.Importance.MEDIUM, SaslConfigs.SASL_MECHANISM_DOC)
+                .define(SaslConfigs.SASL_MECHANISM, ConfigDef.Type.STRING, SaslConfigs.DEFAULT_SASL_MECHANISM, ConfigDef.CompositeValidator.of(new ConfigDef.NonNullValidator(), new ConfigDef.NonEmptyString()), ConfigDef.Importance.MEDIUM, SaslConfigs.SASL_MECHANISM_DOC)
```

Comment:
I don't think we can add this; both `null` and `""` are valid if SASL isn't enabled at all.

Reply:
In fact, even if we don't turn on the sasl mechanism. After adding this non-null and non-empty validator, there is no problem. Because `SaslConfigs.SASL_MECHANISM` has default value `GSSAPI`

Reply:
I'm imagining a situation like this: a user has set `sasl.mechanism` to something strange like `null` (not understanding that this isn't necessary), they haven't configured their consumer/producer to use SASL, then they upgrade to a version of Kafka that includes this change, and things unexpectedly start breaking for them.

I understand that most people don't put garbage in their configs, but I also don't think it's worth the tradeoff here of causing configs that used to be valid to suddenly start failing.

What I believe is more warranted here is context-dependent validation that uses the same core logic that you propose here (i.e., disallowing null and empty values), but only does it if SASL is enabled for the consumer/producer. And, if SASL is disabled and the user has provided a value for this property anyways, it's probably worth emitting a warning letting them know that this value is not being used by the consumer/producer.

Reply:
Hi @C0urante 
On a second thought, SASL modules are highly flexible and customizable.
And this parameter is not a required parameter, which may cause compatibility problems when adding a validator.
So to not make things more complicated, remove the validator for this parameter.
Thanks.